File: docs/perf-comparison/mandelbrot.py
import time

def mandelbrot():
    total = 0
    for y in range(128):
        cy = (y - 64) * 8
        for x in range(128):
            cx = (x - 64) * 8
            zx = 0
            zy = 0
            iter = 0
            while iter < 256:
                zx2 = (zx * zx) >> 8
                zy2 = (zy * zy) >> 8
                if (zx2 + zy2) > 1024:
                    break
                zy = ((zx * zy) >> 7) + cy
                zx = zx2 - zy2 + cx
                iter += 1
            total += iter
    return total

if __name__ == '__main__':
    start = time.perf_counter()
    result = mandelbrot()
    ms = (time.perf_counter() - start) * 1000
    print(f"Python,128,{result},{ms:.1f}")

File: docs/perf-comparison/mandelbrot.cpp
#include <iostream>
#include <chrono>
#include <cstdint>

int64_t mandelbrot() {
    int64_t total = 0;
    for (int y = 0; y < 128; y++) {
        int64_t cy = static_cast<int64_t>(y - 64) * 8;
        for (int x = 0; x < 128; x++) {
            int64_t cx = static_cast<int64_t>(x - 64) * 8;
            int64_t zx = 0, zy = 0;
            int iter = 0;
            while (iter < 256) {
                int64_t zx2 = (zx * zx) >> 8;
                int64_t zy2 = (zy * zy) >> 8;
                if ((zx2 + zy2) > 1024) break;
                zy = ((zx * zy) >> 7) + cy;
                zx = zx2 - zy2 + cx;
                iter++;
            }
            total += iter;
        }
    }
    return total;
}

int main(int argc, char** argv) {
    (void)argc; (void)argv;
    auto start = std::chrono::steady_clock::now();
    int64_t result = mandelbrot();
    auto end = std::chrono::steady_clock::now();
    double us = std::chrono::duration<double, std::micro>(end - start).count();
    std::cout << "C++,128," << result << "," << us << std::endl;
    return 0;
}
